src/sidechaindb.cpp

```diff
@@ -308,179 +360,238 @@ bool SidechainDB::Update(int nHeight, const uint256& hashBlock, const std::vecto
+            wt.nBlocksLeft = ValidSidechains[nSidechain.getint()].GetTau();
+            wt.nWorkScore = 0;
+            wt.wtxid = hashWT;
 
-bool SidechainDB::ApplyStateScript(const CScript& script, bool fJustCheck)
-{
-    if (!HasState())
-        return false;
+            vWT.push_back(wt);
 
-    // Collect the current SCDB status
-    std::vector<std::vector<SidechainWTJoinState>> vState;
-    for (const Sidechain& s : ValidSidechains) {
-        const std::vector<SidechainWTJoinState> vSidechainState = GetState(s.nSidechain);
-        vState.push_back(vSidechainState);
+            // Add to SCDB
+            bool fUpdated = UpdateSCDBIndex(vWT);
+            // TODO handle !fUpdated
+        }
     }
 
-    if (script.size() < 4)
-        return false;
+    // Scan for updated SCDB MT hash and try to update
+    // workscore of WT^(s)
+    // Note: h*(s) and new WT^(s) must be added to SCDB
+    // before this can be done.
+    // Note: Only one MT hash commit is allowed per coinbase
+    std::vector<CScript> vMTHashScript;
+    for (const CTxOut& out : vout) {
+        const CScript& scriptPubKey = out.scriptPubKey;
+        if (scriptPubKey.IsMTHashCommit())
+            vMTHashScript.push_back(scriptPubKey);
+    }
 
-    uint8_t nSidechainIndex = 0;
-    size_t nWTIndex = 0;
-    for (size_t i = 3; i < script.size(); i++) {
-        if (!SidechainNumberValid(nSidechainIndex))
-            return false;
+    if (vMTHashScript.size() == 1) {
+        const CScript& scriptPubKey = vMTHashScript.front();
 
-        // Move on to this sidechain's next WT^
-        if (script[i] == SCOP_WT_DELIM) {
-            nWTIndex++;
-            continue;
+        // Get MT hash from script
+        CScript::const_iterator phash = scriptPubKey.begin() + 6;
+        opcodetype opcode;
+        std::vector<unsigned char> vch;
+        if (scriptPubKey.GetOp(phash, opcode, vch) && vch.size() == 32) {
+            // Try and sync
+            uint256 hashMerkleRoot = uint256(vch);
+            bool fUpdated = UpdateSCDBMatchMT(hashMerkleRoot);
+            // TODO handle !fUpdated
         }
+    }
 
-        // Move on to the next sidechain
-        if (script[i] == SCOP_SC_DELIM) {
-            nWTIndex = 0;
-            nSidechainIndex++;
-            continue;
-        }
+    // Update hashBLockLastSeen
+    hashBlockLastSeen = hashBlock;
 
-        // Check for valid vote type
-        const unsigned char& vote = script[i];
-        if (vote != SCOP_REJECT && vote != SCOP_VERIFY && vote != SCOP_IGNORE)
-            continue;
+    return true;
+}
 
-        if (nSidechainIndex > vState.size())
-            return false;
-        if (nWTIndex > vState[nSidechainIndex].size())
+bool SidechainDB::UpdateSCDBIndex(const std::vector<SidechainWTJoinState>& vNewScores)
+{
+    // First check that sidechain numbers are valid
+    for (const SidechainWTJoinState& s : vNewScores) {
+        if (!SidechainNumberValid(s.nSidechain))
```

Comment:
So if a *single* sidechain id is invalid, we don't want to decrement the work scores for the rest of the sidechains?